### DISCUSSION TITLE
ESIMD: No device of requested type available. -1 (CL_DEVICE_NOT_FOUND)

### DISCUSSION THREAD

**Discussion by xiaoyun2** [intel/llvm#4493] [Q&A]
I'm using dpcpp_compiler 27c632e to compiler and run the offical example [vadd_usm.cpp](https://github.com/intel/llvm-test-suite/blob/intel/SYCL/ESIMD/vadd_usm.cpp).

compiling:
`clang++ -fsycl vadd_usm.cpp`
running:
`SYCL_DEVICE_FILTER=opencl ./a.out`

And I get the following runtime error message:
`terminate called after throwing an instance of 'cl::sycl::runtime_error'
  what():  No device of requested type available. -1 (CL_DEVICE_NOT_FOUND)
Aborted (core dumped)`

```
~: gdb ./a.out
~: (gdb) r
~: (gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff6b13921 in __GI_abort () at abort.c:79
#2  0x00007ffff7ad6957 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ffff7adcae6 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff7adcb21 in std::terminate() ()
   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff7adcd54 in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007ffff71d194e in cl::sycl::device_selector::select_device() const [clone .localalias.4] () from /home/xiaoyun/Downloads/dpcpp_compiler/lib/libsycl.so.5
#7  0x0000000000408485 in cl::sycl::queue::queue(cl::sycl::device_selector const&, std::function<void (cl::sycl::exception_list)> const&, cl::sycl::property_list const&) ()
#8  0x0000000000403a7d in main ()
```

How should I fix this problem, can anyone give some help?

**Comment by alexbatashev** [comment]
@xiaoyun2 please follow [this](https://intel.github.io/llvm-docs/GetStartedGuide.html#install-low-level-runtime) section of Get Started Guide. Alternatively, you can try installing intel-oneapi-runtime-libs from your package manager: https://software.intel.com/content/www/us/en/develop/documentation/installation-guide-for-intel-oneapi-toolkits-linux/top/installation/install-using-package-managers/apt.html

**Reply by xiaoyun2** [reply]
@alexbatashev Thanks for advice, I have tried installing  `intel-oneapi-runtime-libs` from package manager. 
After do this, it shows:
`Supported 'SYCL_DEVICE_FILTER' env var values are 'gpu' and 'host', 'opencl' does not contain such substrings.`
so I run the code by:
`SYCL_DEVICE_FILTER=gpu ./a.out`
It has same runtime error as before.

In the code, `esimd_test::ESIMDSelector{}` is used:
```
queue q(esimd_test::ESIMDSelector{}, esimd_test::createExceptionHandler());
auto dev = q.get_device();
std::cout << "Running on " << dev.get_info<info::device::name>() << "\n";
```

**Reply by alexbatashev** [reply]
@xiaoyun2 could you please show me output of `sycl-ls` (also built as part of compiler toolchain).

UPD: also one important notice: SYCL_DEVICE_FILTER only works correctly with default_selector.

**Reply by xiaoyun2** [reply]
```
~: sycl-ls
[opencl:0] CPU : Intel(R) OpenCL 2.1 [2021.12.6.0.19_160000]
[opencl:0] ACC : Intel(R) FPGA Emulation Platform for OpenCL(TM) 1.2 [2021.12.6.0.19_160000]
[host:0] HOST: SYCL host platform 1.2 [1.2]
```
If I want to running on GPU, what lib or driver should I install?

**Reply by alexbatashev** [reply]
@xiaoyun2 if your system has an Intel GPU, you can follow the instructions from https://github.com/intel/compute-runtime

**Reply by xiaoyun2** [reply]
@alexbatashev Thanks for quick reply, I have installed the compute-runtime from *deb, but it still give same error.
```
:~/Downloads/neo$ sudo dpkg -i *.deb
(Reading database ... 250866 files and directories currently installed.)
Preparing to unpack intel-gmmlib_21.2.1_amd64.deb ...
Unpacking intel-gmmlib (21.2.1) over (21.2.1) ...
Preparing to unpack intel-igc-core_1.0.8365_amd64.deb ...
Unpacking intel-igc-core (1.0.8365) over (1.0.8365) ...
Preparing to unpack intel-igc-opencl_1.0.8365_amd64.deb ...
Unpacking intel-igc-opencl (1.0.8365) over (1.0.8365) ...
Preparing to unpack intel-level-zero-gpu_1.1.20767_amd64.deb ...
Unpacking intel-level-zero-gpu (1.1.20767) over (1.1.20767) ...
Preparing to unpack intel-ocloc_21.34.20767_amd64.deb ...
Unpacking intel-ocloc (21.34.20767) over (21.34.20767) ...
Preparing to unpack intel-opencl_21.34.20767_amd64.deb ...
Unpacking intel-opencl (21.34.20767) over (21.34.20767) ...
Setting up intel-gmmlib (21.2.1) ...
Setting up intel-igc-core (1.0.8365) ...
Setting up intel-igc-opencl (1.0.8365) ...
Setting up intel-level-zero-gpu (1.1.20767) ...
Setting up intel-ocloc (21.34.20767) ...
Setting up intel-opencl (21.34.20767) ...
Processing triggers for libc-bin (2.27-3ubuntu1.2) ...
```

**Reply by alexbatashev** [reply]
@xiaoyun2 are you sure you have a supported Intel GPU? The documentation says, that only Intel HD Graphics 600 or later are supported.

Also, you can check if your OpenCL driver is properly installed with `clinfo` utility (available in apt repository).

**Reply by xiaoyun2** [reply]
Yes, I'm sure I have a supported GPGPU, it's Intel® Iris® Plus Graphics 650. 
But `clinfo` cannot list this device.

**Reply by alexbatashev** [reply]
@xiaoyun2 looks like your issue is GPU-related and is not connected to DPC++. I recommend you asking in https://github.com/intel/compute-runtime repo. One more thing that you may try for self-diagnostic is running clinfo with `LD_DEBUG=all` env variable set to see if the driver is even loaded.

**Comment by DarshanRamakant** [comment]
I am facing similar issue : https://github.com/oneapi-src/oneAPI-samples/issues/1337
